mlir/lib/Analysis/Presburger/Printer.cpp

```diff
@@ -0,0 +1,169 @@
+      os << (i != 0 ? ", " : "") << 's' << i;
+    os << "]";
+  }
+}
+
+/// Prints the constraints of each `FlatAffineConstraints`.
+///
+void PresburgerPrinter::printConstraints(
+    raw_ostream &os,
+    const SmallVectorImpl<FlatAffineConstraints> &flatAffineConstraints) {
+  os << "(";
+  bool fst = true;
+  for (auto &c : flatAffineConstraints) {
+    if (fst)
+      fst = false;
+    else
+      os << " or ";
+    printFlatAffineConstraints(os, c);
+  }
+  os << ")";
+}
+
+/// Prints the constraints of the `FlatAffineConstraints`. Each constraint is
+/// printed separately and the are conjuncted with 'and'.
+///
+void PresburgerPrinter::printFlatAffineConstraints(
+    raw_ostream &os, const FlatAffineConstraints &cs) {
+  unsigned numIds = cs.getNumIds();
+  for (unsigned i = 0, e = cs.getNumEqualities(); i < e; ++i) {
+    if (i != 0)
+      os << " and ";
+    ArrayRef<int64_t> eq = cs.getEquality(i);
+    printExpr(os, eq.take_front(numIds), eq[numIds], cs.getNumDimIds());
+    os << " = 0";
+  }
+
+  if (cs.getNumEqualities() > 0 && cs.getNumInequalities() > 0)
+    os << " and ";
+
+  for (unsigned i = 0, e = cs.getNumInequalities(); i < e; ++i) {
+    if (i != 0)
+      os << " and ";
+    ArrayRef<int64_t> ineq = cs.getInequality(i);
+    printExpr(os, ineq.take_front(numIds), ineq[numIds], cs.getNumDimIds());
+    os << " >= 0";
+  }
+}
+
+/// Prints the coefficient of the i'th variable with an additional '+' or '-' is
+/// first = false. First indicates if this is the first summand of an
+/// expression.
+///
+/// Returns failure if the coefficient value is 0 and therefore is not printed.
+///
+LogicalResult PresburgerPrinter::printCoef(raw_ostream &os, int64_t val,
+                                           bool first) {
+  if (val == 0)
+    return failure();
+
+  if (val > 0) {
+    if (!first) {
+      os << " + ";
+    }
+    if (val > 1)
+      os << val;
+  } else {
+    if (!first) {
+      os << " - ";
+      if (val != -1)
+        os << -val;
+    } else {
+      if (val == -1)
+        os << "-";
+      else
+        os << val;
+    }
+  }
+  return success();
+}
+
+/// Prints the identifier of the i'th variable. The first nDim variables are
+/// dimensions and therefore prefixed with 'd', everything afterwards is a
+/// symbol with prefix 's'.
+///
+void PresburgerPrinter::printVarName(raw_ostream &os, int64_t i,
+                                     unsigned nDim) {
+  if (i < nDim) {
+    os << 'd' << i;
+  } else {
+    os << 's' << (i - nDim);
+  }
+}
+
+/// Prints a constant with an additional '+' or '-' is first = false. First
+/// indicates if this is the first summand of an expression.
+void PresburgerPrinter::printConst(raw_ostream &os, int64_t c, bool first) {
+  if (first) {
+    os << c;
+  } else {
+    if (c > 0)
+      os << " + " << c;
+    else if (c < 0)
+      os << " - " << -c;
+  }
+}
+
+/// Prints a Presburger expression. `coeffs` contains all the coefficients:
```

Comment:
I feel the documentation can be slightly changed to be more specific: this prints an affine expression (Presburger expressions could be piecewise)